File: DFRobet_ColorTemperature.cpp
/*!
 * @file DFRobet_ColorTemperature.cpp
 * @brief This is the implementation file for DFRobet_ColorTemperature
 * @copyright   Copyright (c) 2010 DFRobot Co.Ltd (http://www.dfrobot.com)
 * @license     The MIT License (MIT)
 * @author [TangJie](jie.tang@dfrobot.com)
 * @version  V1.0
 * @date  2024-04-01
 * @url https://github.com/DFRobot/DFRobet_ColorTemperature
 */

#include "DFRobet_ColorTemperature.h"

DFRobet_ColorTemperature::DFRobet_ColorTemperature(TwoWire *pWire)
{
  _pWire = pWire;
  _addr = COLORTEMPERATURE_ADDR;
};

DFRobet_ColorTemperature::DFRobet_ColorTemperature(Stream *s):DFRobot_RTU(s)
{
  _s = s;
  _addr = COLORTEMPERATURE_ADDR;
};

int8_t DFRobet_ColorTemperature::begin(void)
{
  delay(500);
  setTimeoutTimeMs(200);
  
  if(_addr > 0xF7){
    DBG("Invaild Device addr.");
  }
  if(_addr != 0){
    if(!detectDeviceAddress(_addr)){
      DBG("Device addr Error.");
      return -1;
    }
  }else{
    return -1;
  }
  return 0;
}

uint16_t DFRobet_ColorTemperature::readLUX(void)
{
  uint16_t data = 0;
  uint8_t buffer[4];
  readReg(COLORTEMPERATURE_LUX_REG,buffer,2);
  if(_pWire){
    data = buffer[1] << 8 | buffer[0];
  }else{
    data = buffer[0] << 8 | buffer[1];
  }
  
  return data;
}

uint16_t DFRobet_ColorTemperature::readCCT(void)
{
  uint16_t data = 0;
  uint8_t buffer[4];
  readReg(COLORTEMPERATURE_CCT_REG,buffer,2);
  if(_pWire){
    data = buffer[1] << 8 | buffer[0];
  }else{
    data = buffer[0] << 8 | buffer[1];
  }
  return data;
}

float DFRobet_ColorTemperature::readX(void)
{
  float data = 0;
  uint8_t buffer[4];
  readReg(COLORTEMPERATURE_X_REG,buffer,2);
  if(_pWire){
    data = (float)(buffer[1] << 8 | buffer[0]) / 10000;
  }else{
    data = (float)(buffer[0] << 8 | buffer[1]) / 10000;
  }
  
  return data;
}

float DFRobet_ColorTemperature::readY(void)
{
  float data = 0;
  uint8_t buffer[4];
  readReg(COLORTEMPERATURE_Y_REG,buffer,2);
  if(_pWire){
    data = (float)(buffer[1] << 8 | buffer[0]) / 10000;
  }else{
    data = (float)(buffer[0] << 8 | buffer[1]) / 10000;
  }
  return data;
}

bool  DFRobet_ColorTemperature::detectDeviceAddress(uint8_t addr)
{
uint8_t buf[2];
  if(_pWire){
  _pWire->begin();
  }
  readReg(COLORTEMPERATURE_ADDR_REG, buf, 2);
  DBG((buf[0] << 8| buf[1]));
  if(_pWire){
    if(addr == ((buf[1] << 8| buf[0]) & 0xFF))
    return true;
  }else{
    if(addr == ((buf[0] << 8| buf[1]) & 0xFF))
    return true;
  }
  return false;
}

uint8_t DFRobet_ColorTemperature::readReg(uint16_t reg, void *pBuf, uint8_t size)
{
  uint8_t* _pBuf = (uint8_t*)pBuf;
  
    if(pBuf == NULL){
      DBG("data error");
      return 0;
    }
  if(_pWire){
    uint8_t _reg = reg;
    _pWire->begin();
    _pWire->beginTransmission(_addr);
    _pWire->write(_reg);
    _pWire->endTransmission();
    _pWire->requestFrom(_addr, size);
    for(uint8_t i = 0; i < size; i++)
      _pBuf[i] = _pWire->read();
    
    return size;
  }else{
    return readInputRegister(_addr, reg, _pBuf, size);
  }
}

// uint8_t DFRobet_ColorTemperature::writeReg(uint8_t reg, void *pBuf, size_t size)
// {
//   uint8_t *_pBuf = (uint8_t*)pBuf;

//   uint8_t ret = 0;
//   if(_pWire){
//     uint8_t _reg = reg+5;
//     _pWire->beginTransmission(_addr);
//     _pWire->write(_reg);
//     for(uint8_t i = 0; i < size;){
//       uint8_t temp = _pBuf[i];
//       _pBuf[i] = _pBuf[i+1];
//       _pBuf[i+1] = temp;
//       i+=2;
//     }
//     for(size_t i = 0; i < size; i++){
//       _pWire->write(_pBuf[i]);
//     }
//     _pWire->endTransmission();
//   }else{
//     ret = writeHoldingRegister(_addr,reg,_pBuf,size);
//   }
//   return ret;
// }

File: python/raspberry/dfrobet_color_temperature.py
# -*- coding: utf-8 -*
'''!
  @file        dfrobet_color_temperature.py
  @brief       This is the implementation file for DFRobet_ColorTemperature
  @copyright   Copyright (c) 2021 DFRobot Co.Ltd (http://www.dfrobot.com)
  @license     The MIT License (MIT)
  @author      TangJie(jie.tang@dfrobot.com)
  @version     V1.0
  @date        2024-04-01
  @url         https://github.com/DFRobot/DFRobet_ColorTemperature
'''

import serial
import time
import smbus
import os
import math
import RPi.GPIO as GPIO
import math

import modbus_tk
import modbus_tk.defines as cst
from modbus_tk import modbus_rtu

I2C_MODE                  = 0x01
UART_MODE                 = 0x02
DEV_ADDRESS               = 0x32

DEVICE_ADDRESS            = 0x32


class DFRobet_Color_Temperature():
  '''!
    @brief Define DFRobet_ColorTemperature basic class
    @details For driving the environmental sensor
  '''
  def __init__(self ,bus ,baud = 9600):
    if bus != 0:
      self.i2cbus = smbus.SMBus(bus)
      self._uart_i2c = I2C_MODE
    else:
      self.master = modbus_rtu.RtuMaster(serial.Serial(port="/dev/ttyAMA0",baudrate=baud, bytesize=8, parity='N', stopbits=1))
      self.master.set_timeout(1.0)
      self._uart_i2c = UART_MODE

  def _detect_device_address(self):
    '''!
      @brief Get sensor address
      @return  Return sensor address
    '''
    rbuf = self._read_reg(0X02,2)
    if self._uart_i2c == I2C_MODE:
      data = rbuf[1] << 8 | rbuf[0]
    elif self._uart_i2c == UART_MODE:
      data = rbuf[0]
    return data

  def begin(self):
    '''!
      @brief Init SEN0500/SEN0501 sensor
      @return Return init status
      @retval 0  Succeed
      @retval -1 Failed
    '''
    if self._detect_device_address() != DEV_ADDRESS:
      return False
    return True
  
  def read_lux(self):
    '''!
      @brief Get sensor illuminance
      @return Illuminance obtained
    '''
    rbuf = self._read_reg(0X06,2)
    if self._uart_i2c == I2C_MODE:
      data = rbuf[1] << 8 | rbuf[0]
    elif self._uart_i2c == UART_MODE:
      data = rbuf[0]
    return data
  
  def read_cct(self):
    '''!
      @brief Get color temperature data
      @return Color temperature obtained
    '''
    rbuf = self._read_reg(0X07,2)
    if self._uart_i2c == I2C_MODE:
      data = rbuf[1] << 8 | rbuf[0]
    elif self._uart_i2c == UART_MODE:
      data = rbuf[0]
    return data
  
  def read_x(self):
    '''!
      @brief Get color coordinate X
      @return Color coordinate X obtained
    '''
    rbuf = self._read_reg(0X08,2)
    if self._uart_i2c == I2C_MODE:
      data = float(rbuf[1] << 8 | rbuf[0]) / 10000.0
    elif self._uart_i2c == UART_MODE:
      data = float(rbuf[0]) /10000.0
    return data
  
  def read_y(self):
    '''!
      @brief Get color coordinate Y
      @return Color coordinate Y obtained
    '''
    rbuf = self._read_reg(0X09,2)
    if self._uart_i2c == I2C_MODE:
      data = float(rbuf[1] << 8 | rbuf[0]) / 10000.0
    elif self._uart_i2c == UART_MODE:
      data = float(rbuf[0]) / 10000.0
    return data
  
class DFRobet_Color_Temperature_I2C(DFRobet_Color_Temperature):
  '''!
    @brief An example of an i2c interface module
  '''
  def __init__(self ,bus ,addr):
    self._addr = addr
    DFRobet_Color_Temperature.__init__(self,bus,0)   
    
  
  def _read_reg(self, reg_addr ,length):
    '''!
      @brief read the data from the register
      @param reg register address
      @param value read data
    '''
    try:
      rslt = self.i2cbus.read_i2c_block_data(self._addr ,reg_addr , length)
    except:
      rslt = -1
    return rslt    


class DFRobet_Color_Temperature_UART(DFRobet_Color_Temperature):
  '''!
    @brief An example of an UART interface module
  '''
  def __init__(self ,baud, addr):
    self._baud = baud
    self._addr = addr
    try:
      DFRobet_Color_Temperature.__init__(self,0,self._baud)
    except:
      print ("plese get root!")
   
  
  def _read_reg(self, reg_addr ,length):
    '''!
      @brief Read data from the sensor
    '''
    return list(self.master.execute(self._addr, cst.READ_INPUT_REGISTERS, reg_addr, length))
    